server/src/test/java/org/elasticsearch/index/shard/IndexShardTests.java

```diff
@@ -4559,6 +4559,33 @@ public void testResetEngine() throws Exception {
         closeShard(shard, false);
     }
 
+    public void testResetEngine() throws Exception {
+        var newEngineCreated = new CountDownLatch(2);
+        var indexShard = newStartedShard(true, Settings.EMPTY, config -> {
+            try {
+                return new ReadOnlyEngine(config, null, null, true, Function.identity(), true, true) {
+                    @Override
+                    public void prepareForEngineReset() throws IOException {
+                        ;
+                    }
+                };
+            } finally {
+                newEngineCreated.countDown();
+            }
+        });
+        var newEngineNotification = new CountDownLatch(1);
+        indexShard.waitForEngineOrClosedShard(ActionListener.running(newEngineNotification::countDown));
```

Comment:
shouldn't we reset this listener too? i.e. a caller can call `waitForPrimaryTermAndGeneration` while the engine is being transitioned to the new one and proceed with the call and use the old one? maybe it's not important but it would be good to be consistent.

Reply:
Sounds good